[V1] implement tree sampler for draft token acceptance

TheEpicDolphin opened the pull request:
# Purpose
Continuing with my work in this PR: #20401 , where I added support for drafting a tree of speculative tokens, that are then validated by the target model. In this PR, I add the class that performs rejection sampling for those draft tokens, so that they conform to the target model's output distribution. This class is based off of RejectionSampler, but with some key differences necessary to support a tree structure for drafted tokens. I added some tests for this new class to verify it's correctness.

In order to make it work with tree-drafted tokens, I also had to "remap" the KV cache immediately after rejection sampling. This is handled by the new `_remap_kv_cache` method I added. It basically takes the KVs for the `N` accepted branch of tokens, and copies them over into the physical KV cache memory corresponding to the first `N` contiguous paged KV slots for the draft request.

In addition, I also made some refactors to the tree attention parameters to improve readability and performance. I created a new class called `TreeDrafterParams` which is created during the `SpeculativeConfig` initialization, and precomputes several properties from the spec token tree so that other tree-attention systems can use it (without re-computing themselves). Examples: attention mask, children per level, etc.

With this PR, we now have a **fully functional** tree speculative decoding implementation in V1!

# Test Plan

## Automated Testing
New tree rejection sampler tests:
```
(py312conda) bash-5.1$ pytest tests/v1/sample/test_tree_rejection_sampler.py
=============================================================================== test session starts ===============================================================================
platform linux -- Python 3.12.9, pytest-8.4.1, pluggy-1.6.0
rootdir: /data/users/gdelfin/gitrepos/vllm
configfile: pyproject.toml
plugins: anyio-4.9.0, asyncio-1.1.0
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 9 items                                                                                                                                                                 

tests/v1/sample/test_tree_rejection_sampler.py .........                                                                                                                    [100%]

================================================================================ 9 passed in 6.51s ================================================================================
```

Eagle tree proposer test:
```
(py312conda) bash-5.1$ pytest tests/v1/spec_decode/test_eagle.py -k test_propose_tree
=============================================================================== test session starts ===============================================================================
platform linux -- Python 3.12.9, pytest-8.4.1, pluggy-1.6.0
rootdir: /data/users/gdelfin/gitrepos/vllm
configfile: pyproject.toml
plugins: anyio-4.9.0, asyncio-1.1.0
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 47 items / 43 deselected / 4 selected                                                                                                                                   

tests/v1/spec_decode/test_eagle.py ....                                                                                                                                     [100%]

======================================================================== 4 passed, 43 deselected in 10.82s ========================================================================
```

Spec decode e2e test:

```
(py312conda) bash-5.1$ pytest tests/v1/e2e/test_spec_decode.py -k test_tree_eagle_correctness
=============================================================================================================================== test session starts ===============================================================================================================================
platform linux -- Python 3.12.9, pytest-8.4.1, pluggy-1.6.0
rootdir: /data/users/gdelfin/gitrepos/vllm
configfile: pyproject.toml
plugins: anyio-4.9.0, asyncio-1.1.0
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 21 items / 13 deselected / 8 selected                                                                                                                                                                                                                                   

tests/v1/e2e/test_spec_decode.py ........                                                                                                                                                                                                                                   [100%]

================================================================================================================================ warnings summary =================================================================================================================================
tests/v1/e2e/test_spec_decode.py: 16 warnings
  /home/gdelfin/.conda/envs/py312conda/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=2475023) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================================================ 8 passed, 13 deselected, 16 warnings in 382.20s (0:06:22) ============================================================================================================
```

## Manual Testing

I tested manually with the following tree of draft tokens:
```
ROOT
├── 0
│  ├── 0
│  └── 1
│  └── 2
└── 1
   ├── 0
   └── 1
   └── 2
```

**Server**
```
export VLLM_TORCH_PROFILER_DIR=~/traces/vllm
export LLAMA_MODEL=meta-llama/Llama-3.1-8B-Instruct
export DRAFT_MODEL=yuhuili/EAGLE-LLaMA3.1-Instruct-8B
export VLLM_USE_V1=1
export VLLM_ATTENTION_BACKEND=TREE_ATTN
export SPEC_DEC_CONFIG='{"method": "eagle", "model": "'$DRAFT_MODEL'", "num_speculative_tokens": 8, "draft_tensor_parallel_size": 1, "max_model_len": 2048, "speculative_token_tree": "[(0,), (1,), (0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2)]"}'
vllm serve $LLAMA_MODEL --disable-log-requests --tensor-parallel-size=1 --max-num-seqs=64 --max-model-len=32768 --no-enable-prefix-caching --speculative-config="$SPEC_DEC_CONFIG" 2>&1 | tee ~/server_logs/vllm_server.log
```

**Client**
```
from openai import OpenAI
client = OpenAI(base_url="http://localhost:8000/v1", api_key="EMPTY")
response = client.chat.completions.create(model="meta-llama/Llama-3.1-8B-Instruct", messages=[{"role": "system", "content": "You are a helpful assistant."}, {"role": "user", "content": "Explain the theory of relativity in simple terms."}],temperature=0.0)
print(response.choices[0].message.content)
```

**Response**
```
The theory of relativity, developed by Albert Einstein, is a fundamental concept in physics that explains how space and time are connected. It's a bit complex, but I'll try to break it down in simple terms.

**Special Relativity (1905)**

Imagine you're on a train, and you throw a ball straight up in the air. From your perspective on the train, the ball goes up and comes down in a straight line. Now, imagine someone is standing outside the train, watching you throw the ball. From their perspective, the ball doesn't just go up and down – it also moves forward, because the train is moving really fast.

Einstein said that how we measure time and space depends on how fast we're moving and where we are. If you're on the train, time and space seem normal. But if you're standing outside, watching the train, time and space seem different. This is because time and space are connected, and they can appear to change depending on your frame of reference.

**Time Dilation**

Here's a key idea: time can appear to slow down or speed up depending on how fast you're moving. If you were to travel close to the speed of light, time would appear to slow down for you relative to someone who is standing still. This is called time dilation.

**General Relativity (1915)**

Imagine you're standing on a trampoline. If you put a heavy object, like a bowling ball, on the trampoline, it will warp and curve, creating a dent. That's kind of like what gravity does to space and time. According to general relativity, massive objects like planets and stars warp the fabric of space and time, creating gravity.

**Key Takeaways**

1. **Time and space are connected**: How we measure time and space depends on how fast we're moving and where we are.
2. **Time can appear to slow down or speed up**: Time dilation occurs when you're moving at high speeds or in strong gravitational fields.
3. **Gravity warps space and time**: Massive objects create curvatures in space and time, which we experience as gravity.

The theory of relativity revolutionized our understanding of the universe and has had a profound impact on modern physics and astronomy.
```
The response is valid and aligns with what the LLM would output if running with standard decoding or spec decoding with FA3.

# Benchmarks
I benchmarked using the philschmid/mt-bench and likaixin/InstructCoder datasets. Below are the commands I used:
**Server**
```
export VLLM_TORCH_PROFILER_DIR=~/traces/vllm
export LLAMA_MODEL=meta-llama/Llama-3.1-8B-Instruct
export DRAFT_MODEL=yuhuili/EAGLE-LLaMA3.1-Instruct-8B
export VLLM_USE_V1=1
export VLLM_ATTENTION_BACKEND=TREE_ATTN
export SPEC_DEC_CONFIG='{"method": "eagle", "model": "'$DRAFT_MODEL'", "num_speculative_tokens": 8, "draft_tensor_parallel_size": 1, "max_model_len": 2048, "speculative_token_tree": "[(0,), (1,), (0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2)]"}'
vllm serve $LLAMA_MODEL --disable-log-requests --tensor-parallel-size=1 --max-num-seqs=64 --max-model-len=32768 --no-enable-prefix-caching --speculative-config="$SPEC_DEC_CONFIG" 2>&1 | tee ~/server_logs/vllm_server.log
```

**Client**
```
vllm bench serve --model $LLAMA_MODEL --tokenizer $LLAMA_MODEL --host 0.0.0.0 --dataset-name hf --dataset-path <dataset> --ignore-eos --request-rate inf --max-concurrency <batch_size>
```

Using both datasets and batch sizes 1 and 8, I compared the serving benchmark results for 2 spec tokens (baseline), and a 2 => 3 spec token tree that looks like this:
```
ROOT
├── 0
│  ├── 0
│  └── 1
│  └── 2
└── 1
   ├── 0
   └── 1
   └── 2
```
For the baseline, I decided to use the flash attention 3 backend, since it's the only other one that supports spec decoding in V1. Below is a short summary of the results:

**mt-bench**
| Metric                          | BS=1, FA3 | BS=1, Tree | BS=8, FA3 | BS=8, Tree | BS=16, FA3 | BS=16, Tree |
|---------------------------------|-----------|------------|-----------|------------|------------|-------------|
| Duration (s)                     | 1999.14   | 1990.13    | 269.42    | 265.15     | 118.45     | 131.05      |
| Output Token Throughput (tok/s) | 128.06    | 128.63     | 884.61    | 902.72     | 1813.36    | 1640.95     |
| Mean Acceptance Length           | 2.10      | 2.37       | 2.03      | 2.37       | 2.04       | 2.38        |

In all cases, tree spec decoding **increases** the mean acceptance length. In the mt-bench dataset, tree spec decoding results in a **higher output token throughput** for batch sizes 1 and 8. As batch size gets larger, the benefit of more accepted tokens per decode is not enough to compensate for the added cost during forward attention, and tree spec decoding performs worse. It's important to note that the FA3 kernel is much faster than triton attention, which is used in the Tree Attention backend (with a query-on-query attention bias).

To isolate what we are gaining from tree drafts, I also compared chains vs tree draft structures using the same Tree Attention backend. This eliminates difference in kernel

**mt-bench**
| Metric                          | BS=1, Chain | BS=1, Tree | BS=8, Chain | BS=8, Tree | BS=16, Chain | BS=16, Tree |
|---------------------------------|-------------|------------|-------------|------------|-------------|------------|
| Duration (s)                     | 2240.21           | 1990.13    | 280.96      | 265.15     | 136.86      | 131.05     |
| Output Token Throughput (tok/s) | 114.27       | 128.63     | 847.38      | 902.72     | 1567.5      | 1640.95    |
| Mean Acceptance Length           | 2.05        | 2.37       | 2.06       | 2.37       | 2.05        | 2.38       |


Spec decoding with the tree of drafts performs better than with a chain of 2 drafts for batch sizes 1, 8, and 16. I did not test further with larger batch sizes, but I assume that gain becomes less pronounced.

A detailed breakdown of the benchmark results can be found in [this spreadsheet](https://docs.google.com/spreadsheets/d/1YnOJ2NsUffRjCj0egS_ycPpkeYwpFp4PpH1BBFFaAMs/edit?usp=sharing).

# Next Steps
- Use a different attention kernel because triton attention is slow compared to FA3. From benchmarking, each triton attention forward pass is ~35% slower than the FA3 forward pass. FA2+tree attention mask (https://github.com/vllm-project/flash-attention/pull/81) is a good candidate!
- As @sgrigory pointed out in the comments, this implementation is currently using greedy acceptance, which differs from the probabilistic approach used in EAGLE. The latter should increase the avg # of acceptances and improve performance even further.

TheEpicDolphin committed:
[V1] implement tree sampler for draft token acceptance

Signed-off-by: Giancarlo Delfin <gdelfin@meta.com>

TheEpicDolphin committed:
[V1] implement tree sampler for draft token acceptance

Signed-off-by: Giancarlo Delfin <gdelfin@meta.com>

sgrigory commented on tests/v1/sample/test_tree_rejection_sampler.py:
```diff
@@ -0,0 +1,371 @@
+
+
+def test_deep_tree_full_acceptance():
+    draft_tree = [
```
Is there an assumption that all nodes on a given level of a tree have the same number of children? If yes, can we assert this somewhere? If no, could we also test "pruned" trees

TheEpicDolphin replied:
Correct, i'm only supporting regular trees right now. That is, each node at a given level has the same # of children. Will add an assert for that!

TheEpicDolphin replied:
Added!

sgrigory commented on vllm/v1/sample/tree_rejection_sampler.py:
```diff
@@ -0,0 +1,232 @@
+        output_probs = logits.softmax(dim=-1, dtype=torch.float32)
+        return output_probs.view(batch_size, -1, vocab_size)
+
+    def rejection_sample(self, target_probs: torch.Tensor,
```
This looks like greedy acceptance criterion, which will in general have lower acceptance rate than the probabilistic/rejection sampling-style algo used by EAGLE. IIRC chain spec decoding in vLLM supports the probabilistic acceptance. Do you also plan to implement it for trees? You can have a look at Algorithm 1 in https://arxiv.org/pdf/2401.15077.

wangjiahe0915 commented:
Which version of vLLM are you developed based on?

TheEpicDolphin commented on vllm/v1/sample/rejection_sampler.py:
```diff
@@ -82,6 +82,7 @@ def forward(
             output_token_ids (torch.Tensor):
                 A tensor containing the final output token IDs.
         '''
+        assert bonus_token_ids is not None
```
Addressed in my new approach!

TheEpicDolphin commented:
>Which version of vLLM are you developed based on?
This is on the V1 engine of vLLM spec decoding

TheEpicDolphin commented:
@LucasWilkinson Added to both test plan and spreadsheet! Turns out Tree attention performs better on mt-bench for BS=1, but not for larger batch sizes.

TheEpicDolphin commented on vllm/v1/sample/tree_rejection_sampler.py:
```diff
@@ -0,0 +1,232 @@
+
+        # Get tree depth (# levels) and width (# drafts at last level).
+        self.tree_depth = len(self.cu_drafts_per_level)
+        self.tree_width = self.cu_drafts_per_level[
```
Yeah, tree rejection sampling does not handle the case when `num_speculative_tokens = 0` (just root node). I added an assert to check that there is always at least 1 speculative token when tree attention is enabled!

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3274,3 +3313,107 @@ def _to_list(self, sampled_token_ids: torch.Tensor) -> list[list[int]]:
+        layers = get_layers_from_vllm_config(self.vllm_config, Attention)
+
+        # Get attention metadata for a layer. The current assumption is
+        # that all attention layers share the same KV cache slot mapping.
```
we should assert that there is only one kv-cache group to ensure this is the case; this won't hold true for models that use the hybrid kv-cache (e.g. llama4 with hybrid kv for long context)

cc @luccafong @heheda12345

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1845,6 +1860,21 @@ def propose_draft_token_ids(
                     self.drafter.prepare_inputs(
                     common_attn_metadata, num_rejected_tokens_cpu)
 
+                if remapped_token_indices is not None:
```
nit: is there a way move this logic into the `TreeRejectionSampler` or the proposer? theres generally been complaints that theres so many features in the gpu model runner right now its hard to understand the flow without understanding all the different features so if possible we'd like to move as much as possible from gpu model runner to more feature specific modules (not blocking but just something we strive for; and have be struggling with)

LucasWilkinson commented on vllm/v1/sample/tree_rejection_sampler.py:
```diff
@@ -0,0 +1,275 @@
+            device=tree_mask.device,
+            dtype=torch.int32)
+        parents_end = 0
+        for level in range(self.tree_depth - 1):
```
Theres alot indexing in this function; I think it would be very helpful have a working example in the comments like whats throughout: https://github.com/vllm-project/vllm/blob/2e5d21378db8226b04bb76f3e7119a9ddd0239df/vllm/v1/attention/backends/utils.py#L382-L566

(another example: https://github.com/vllm-project/vllm/blob/2e5d21378db8226b04bb76f3e7119a9ddd0239df/vllm/v1/worker/gpu_model_runner.py#L713-L719)

LucasWilkinson commented on vllm/v1/attention/backends/tree_attn.py:
```diff
@@ -167,21 +167,16 @@ def __init__(
+            # Spec decoding.
+            tree_attn_mask = torch.tensor(tree_drafter_params.attn_mask, device=device)
+            self.tree_attn_bias = torch.where(tree_attn_mask, 0, -torch.inf)
+        self.__class__.reorder_batch_threshold = self.tree_attn_bias.shape[0]
```
what if there is prefill or prefill-chunk that is `<= self.tree_attn_bias.shape[0]` will this get treated as a decode? i.e. will have the `tree_attn_bias` applied? sorry `split_decodes_and_prefills` is a bit of misnormer since it really just splits the batch into `q_len <= decode_threshold` and `q_len > decode_threshold` but short prefills could have a `q_len <= decode_threshold` and this function will treat them the same (this is an artifact of V1 not distinguishing between prefill and decodes)

TheEpicDolphin replied:
Yep, that's a pre-existing limitation of this system. If prefill length <= self.tree_attn_bias.shape[0], it'll be misclassified as a decode, and the tree mask will be incorrectly applied.

I plan to follow up with a fix to that in a future PR. Don't wanna clutter this PR with anymore stuff, so I'll just leave TODO here. Thanks for the callout!

LucasWilkinson reviewed (commented):
overall looks pretty good! thanks for the contribution! left a few comments; i'd love to see if there's a way to move more stuff out of the gpu model runner but I know that can be challenging. Would be helpful if @heheda12345 and @luccafong can also take a look :+1:

LucasWilkinson reviewed (dismissed):
Thanks for the clean-up! Overall looks good to me; we are potentially freezing gpu model runner changes (since its gotten too bloated) until @WoosukKwon can complete his refactor; ill follow-up with him when are clear to merge :+1:

TheEpicDolphin commented:
@LucasWilkinson , @WoosukKwon , i refactored so that there are even fewer changes in gpu model runner. Basically, i moved the logic of fetching bonus and target logits into `RejectionSampler`/`TreeRejectionSampler`. Please let me know what you think

TheEpicDolphin commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1839,31 +1870,15 @@ def _sample(
-            # logits tensor. This means any in-place operations on bonus_logits
-            # won't affect the original logits tensor.
             assert logits is not None
-            bonus_logits = logits[spec_decode_metadata.bonus_logits_indices]
-            sampler_output = self.sampler(
-                logits=bonus_logits,
-                sampling_metadata=sampling_metadata,
-            )
-            bonus_token_ids = sampler_output.sampled_token_ids
-
-            # Just like `bonus_logits`, `target_logits` is a new tensor with
-            # separate storage from the original `logits` tensor. Therefore,
-            # it is safe to update `target_logits` in place.
-            target_logits = logits[spec_decode_metadata.target_logits_indices]
```
This was moved into `RejectionSampler.forward`.
